### ISSUE TITLE
Aporeto external access inconsistencies 

### ISSUE THREAD

**Issue by stewartshea** [BCDevOps/platform-services#103]
During some periods of time, policy operates as it should and pods can communicate externally as well as with the appropriate internal resources. Other times it just bails out; this last occurrence of this was functioning better last night and with no changes stopped this morning: 

- Failed resolution (with any other specific policies disabled, but egress explicitly allowed) 
```
sh-4.2$ curl www.google.ca
curl: (7) Failed to connect to 2607:f8b0:400a:804::2003: Network is unreachable
```

- Remove agents
```
oc label nodes ociopf-t-311.dmz ociopf-t-312.dmz ociopf-t-313.dmz ociopf-t-321.dmz ociopf-t-322.dmz aporeto-enforcerd-
```

- Test resolution
```
sh-4.2$ curl www.google.ca
<!doctype html><html itemscope=""
```

- Add Agents
```
oc label nodes ociopf-t-311.dmz ociopf-t-312.dmz ociopf-t-313.dmz ociopf-t-321.dmz ociopf-t-322.dmz aporeto-enforcerd=true
```

- Test again (both with google.ca and github)
```
sh-4.2$ curl www.google.ca
<!doctype html><html itemscope=""
...

sh-4.2$ curl www.github.com
curl: (7) Failed connect to www.github.com:80; Operation now in progress
```

- Test again after 5-10 minutes
```
sh-4.2$ curl www.google.ca
curl: (7) Failed to connect to 2607:f8b0:400a:800::2003: Network is unreachable
sh-4.2$
```

- Delete pod and test again

```
sh-4.2$ curl www.google.ca
<!doctype html><html itemscope=""
...

sh-4.2$ curl www.github.com
curl: (7) Failed connect to www.github.com:80; Operation now in progress
``` 

Why oh why?

**Comment by mheese** [comment]
When seeing issues like this in the past with Aporeto (sporadic connectivity problems), it has _always_ been a problem with connectivity issues to the DNS servers. Let's go over it in the afternoon together.

Can you try and see if a curl to an IP address always works? E.g. you should see a cgit listing of the Linux kernel source repos here:

```
curl -k https://147.75.56.137/
```

**Comment by stewartshea** [comment]
I have tested this without success either; 
- Curl external IP address
```
sh-4.2$ curl www.google.ca
^C
sh-4.2$ curl -k https://147.75.56.137/

curl: (7) Failed connect to 147.75.56.137:443; Connection timed out
```

- Remove agents & re0add agents
```

sh-4.2$ curl -k https://147.75.56.137/
...
</table></div> <!-- class=content -->
<div class='footer'>generated by <a href='https://git.zx2c4.com/cgit/about/'>cgit 1.2-0.3.lf.el7</a> (<a href='https://git-scm.com/'>git 2.18.0</a>) at 2019-05-02 16:54:47 +0000</div>
</div> <!-- id=cgit -->
</body>
</html>
sh-4.2$
sh-4.2$
sh-4.2$
sh-4.2$
sh-4.2$ curl -k https://147.75.56.137/

```

In addition, we don't see any of the blocked flows: 
![image](https://user-images.githubusercontent.com/18314168/57092929-6f1a6700-6cc1-11e9-8758-a879c3cd53b1.png)

**Comment by mheese** [comment]
This will require some intense debugging I fear.

Because think about it this way: if the connections time out and we don't see any rejected/blocked flows, that would usually mean that they are not blocked by Aporeto.

Do you have anything else running on the underlying hosts that might get in the way for Aporeto? Any other RHEL iptables-based service might get in the way (like firewalld or iptables services). Those are usually only getting in the way for protecting Linux processes though.

**Comment by stewartshea** [comment]
If I remove Aporeto, there are no connectivity issues. I'm not sure what you mean about iptables or firewalld since iptables has to be running in a kube world...

**Comment by mheese** [comment]
I'm referring to the RHEL iptables and firewalld systemd services.

**Comment by stewartshea** [comment]
The only thing in the logs that might be of interest is this line: 
```
ERROR	Failed to update conntrack table	{"error": "Netlink Returned errror -2"}
```

**Comment by stewartshea** [comment]
```

__                            _
--
  | ___ _ __  / _\| ___  _ __ ___ ___ _ __ __\| \|
  | / _ \ '_ \\| \|_ / _ \\| '__/ __/ _ \ '__/ _' \|
  | \|  __/ \| \| \|  _\| (_) \| \| \| (_\|  __/ \| \| (_\| \|
  | \___\|_\| \|_\|_\|  \___/\|_\|  \___\___\|_\|  \__,_\|
  |  
  | Aporeto Enforcer Agent
  | v1.739.16 - ded76cd44d3ad562766838b2614277b5f02a9d69
  | _______________________________________________________________
  |  
  | INFO	Probing cloud providers...
  | INFO	Using generic metadata extractor
  | INFO	Using app credentials
  | INFO	Service	{"api": "https://api.console.aporeto.com "}
  | INFO	Looking for pre existing enforcer	{"namespace": "/bcgov-devex/lab", "machine-id": "4203F85A-72DB-D247-C1E9-39D0256E3F07"}
  | INFO	Enforcer configured in backend	{"enforcerdID": "5cc8a35687450d0001f222dc", "namespace": "/bcgov-devex/lab"}
  | ERROR	Failed to increase resource limit	{"error": "Unable to get current resource limit operation not permitted"}
  | INFO	Update check enabled	{"manifest": "https://download.aporeto.com/manifests/prod-manifest.json ", "currentVersion": "v1.739.16"}
  | ERROR: logging before flag.Parse: W0502 22:03:44.280535   22622 client_config.go:533] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
  | ERROR	Unable to start the audit client. Audit functionality disabled. Ensure auditd daemon is not running	{"error": "Cannot configure backlog limit in the kernel: operation not permitted"}
  | WARN	dropping tag.	{"error": "error 422 (elemental): Validation Error: Data 'annotation.kubernetes.io/limit-ranger=LimitRanger plugin set: cpu, memory request for container enforcerd; cpu, memory limit for container enforcerd' of attribute 'value' must contain at least one '=' symbol separating two valid words."}
  | WARN	dropping tag. value longer than 128	{"tagvalue": ""}
  | WARN	dropping tag. value longer than 128	{"tagvalue": ""}
  | ERROR	Failed to update conntrack table	{"error": "Netlink Returned errror -2"}
  | ERROR	Failed to update conntrack table	{"error": "Netlink Returned errror -2"}
  | ERROR	Failed to update conntrack table	{"error": "Netlink Returned errror -2"}


```

**Comment by stewartshea** [comment]
Removed Sysdig agent to test and validate if there were any conflicts; no change.

**Comment by stewartshea** [comment]
Dropping in the slack comments that we have here --> this is leading us to a good place! 
```
so here are my assumptions:
- we're using the redhat/openshift-multitenant SDN
- there is no other network provider installed for let's say network policies (like Calico or something)
- *we have no issues curling other PUs* ... this one is critical
- we see intermittent (could be 50% and more) outbound connectivity issues to "external networks"
- we have the namespace excluded in the enforcer profile for the namespace where the enforcer is running
now for the hunch
in the enforcer profile, we still have the possibility to set the networks that are "managed" by Aporeto ... this basically means: "in all these CIDRs, I expect to see Aporeto PUs"
that also means that everything besides those CIDRs are always going to be considered CIDRs that can only belong to "external networks"
so traffic to non-managed Aporeto networks will always be treated as external traffic
why is this important?
The requirement for setting managed Aporeto networks was pretty much superfluous very early on ... because we introduced auto-detection for those
it means that we're going to probe if a connection belongs to us or not
now if on the way outbound there is a firewall/router/proxy (whatever) that modifies the TCP connections, then that probing will fail
if one sets managed networks in the enforcer profile, there is never going to be auto-detection done
*TL;DR* - let's add our cluster network and service network to the managed TCP networks in the enforcer profile
we've apparently seen this issue not too long ago with a customer - also in a large enterprise network - 

```

**Comment by stewartshea** [comment]
Updated the enforcer profiles as directed and getting a stable experience. Docs to be updated soon, monitoring to occur, and then we can close this out once all of that is done.

**Comment by stewartshea** [comment]
just need to validate that this is no longer an issue
